controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -422,12 +436,59 @@ public void refresh() {
 
     @Override
     CompletableFuture<Void> createCompletedTxEntry(final UUID txId, final TxnStatus complete, final long timestamp) {
-        final String completedTxPath = getCompletedTxPath(txId.toString());
-        return store.createZNodeIfNotExist(completedTxPath,
-                new CompletedTxnRecord(timestamp, complete).toByteArray())
-                .whenComplete((r, e) -> cache.invalidateCache(completedTxPath));
+        String root = String.format(STREAM_COMPLETED_TX_BATCH_PATH, currentBatchSupplier.getLatestBatch(), getScope(), getName());
+        String path = ZKPaths.makePath(root, txId.toString());
+
+        CompletableFuture<Void> future1 = store.createZNodeIfNotExist(path,
+                        new CompletedTxnRecord(timestamp, complete).toByteArray())
+                        .whenComplete((r, e) -> cache.invalidateCache(completedTxPath));
+
+        // region Backward Compatibility
+        // TODO 2755 retire code in this region 
+        // We will continue to write to old path for backward compatibility.
+        // This should eventually be removed.
+        if (!Config.DISABLE_COMPLETED_TXN_BACKWARD_COMPATIBILITY) {
```

Comment:
To me this upgrade path is not clear. 
* When can a user turn on this flag?
* When can we remove this flag?

Reply:
I think the plan should be add this flag in 0.4.0 and remove it in 0.5.0. Does it make sense?

Reply:
basically it should be removed in the version where we say disallow upgrade path from 0.3.0.